src/game/collide.c

```diff
@@ -937,6 +961,93 @@ int32_t Move3DPosTo3DPos(
         && srcpos->y_rot == destpos->y_rot && srcpos->z_rot == destpos->z_rot;
 }
 
+int32_t Move3DPosTo3DPosAnim(
+    PHD_3DPOS *srcpos, PHD_3DPOS *destpos, int32_t velocity, int16_t rotation,
+    ITEM_INFO *lara_item)
+{
+    int32_t dx = destpos->x - srcpos->x;
+    int32_t dy = destpos->y - srcpos->y;
+    int32_t dz = destpos->z - srcpos->z;
+    int32_t distance = phd_sqrt(SQUARE(dx) + SQUARE(dy) + SQUARE(dz));
+    if (velocity >= distance) {
+        srcpos->x = destpos->x;
+        srcpos->y = destpos->y;
+        srcpos->z = destpos->z;
+    } else {
+        int32_t shift = PHD_ONE * velocity / distance;
+        srcpos->x += shift * dx >> 16;
+        srcpos->y += shift * dy >> 16;
+        srcpos->z += shift * dz >> 16;
+    }
+
+    if (!g_Lara.interact_target.is_moving) {
+        if (g_Lara.water_status != LWS_UNDERWATER) {
+            const int16_t stepToAninNum[4] = {
+                AA_SIDESTEP_LEFT,
+                AA_WALK_FORWARD,
+                AA_SIDESTEP_RIGHT,
+                AA_WALK_BACK,
+            };
+            const int16_t stepToAnimState[4] = {
+                AS_STEPLEFT,
+                AS_WALK,
+                AS_STEPRIGHT,
+                AS_BACK,
+            };
+
+            int32_t angle =
+                (PHD_ONE
+                 - phd_atan(srcpos->x - destpos->x, srcpos->z - destpos->z))
+                % PHD_ONE;
+            uint32_t quadrant =
+                ((((uint32_t)(angle + PHD_45) >> W2V_SHIFT)
+                  - ((uint16_t)(destpos->y_rot + PHD_45) >> W2V_SHIFT))
+                 & 0x3);
```

Comment:
This complex expression should be extracted to a static function.

Also I think it can be simplified to
`((angle - destpos->y_rot) >> W2V_SHIFT) & 3` because `PHD_45` cancel themselves out and the W2V shift can be applied after we're done calculating the final angle. I also think this should work without any casts.

Reply:
Ok. I tried to use the TR5 code directly. It's just a long one liner which is why I split it up.

`((((ulong)(mGetAngle(dest->x_pos, dest->z_pos, pos->x_pos, pos->z_pos) + 8192) >> 14) - ((ushort)(dest->y_rot + 8192) >> 14)) & 0x3)`

Reply:
After some testing I think this formula needs to be changed back. Comparing formulas with angle = 58000 and y_rot = -31000:
```
rr:   ((58000 - -31000 ) >> 14) & 3 = 1
tr5: (((58000 + 8192) >> 14) - ((-31000 + 8192) >> 14)) & 0x3 = (4) - (-2) & 0x3 = 2
```

Reply:
OK but please put it in a function then.

Reply:
Or maybe not. The parent function has gotten small enough to keep it inlined.